File: app/src/org/commcare/android/database/connect/models/PersonalIdCredential.java
package org.commcare.android.database.connect.models;

import org.commcare.android.storage.framework.Persisted;
import org.commcare.models.framework.Persisting;
import org.commcare.modern.database.Table;
import org.commcare.modern.models.MetaField;
import org.javarosa.core.services.Logger;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

/**
 * DB model for personal_id_credential table
 */
@Table(PersonalIdCredential.STORAGE_KEY)
public class PersonalIdCredential extends Persisted implements Serializable {

    public static final String STORAGE_KEY = "personal_id_credential";

    public static final String META_APP_NAME = "app_name";
    public static final String META_SLUG = "slug";
    public static final String META_TYPE = "type";
    public static final String META_ISSUED_DATE = "issued_date";
    public static final String META_TITLE = "title";
    public static final String META_CREDENTIAL = "credential";

    // Static list to keep track of corrupt entries
    private static final List<PersonalIdCredential> corruptCredentialItems = new ArrayList<>();

    @Persisting(1)
    @MetaField(META_APP_NAME)
    private String appName;

    @Persisting(2)
    @MetaField(META_SLUG)
    private String slug;

    @Persisting(3)
    @MetaField(META_TYPE)
    private String type;

    @Persisting(4)
    @MetaField(META_ISSUED_DATE)
    private String issuedDate;

    @Persisting(5)
    @MetaField(META_TITLE)
    private String title;

    @Persisting(6)
    @MetaField(META_CREDENTIAL)
    private String credential;

    // Constructors
    public PersonalIdCredential() {
    }

    public PersonalIdCredential(String appName, String slug, String type,
                                String issuedDate, String title, String credential) {
        this.appName = appName;
        this.slug = slug;
        this.type = type;
        this.issuedDate = issuedDate;
        this.title = title;
        this.credential = credential;
    }

    // Getters
    public String getAppName() {
        return appName;
    }

    public String getSlug() {
        return slug;
    }

    public String getType() {
        return type;
    }

    public String getIssuedDate() {
        return issuedDate;
    }

    public String getTitle() {
        return title;
    }

    public String getCredential() {
        return credential;
    }

    // Setters
    public void setAppName(String appName) {
        this.appName = appName;
    }

    public void setSlug(String slug) {
        this.slug = slug;
    }

    public void setType(String type) {
        this.type = type;
    }

    public void setIssuedDate(String issuedDate) {
        this.issuedDate = issuedDate;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public void setCredential(String credential) {
        this.credential = credential;
    }

    /**
     * Parses a JSON array into a list of PersonalIdCredential objects.
     * Any corrupt entries are logged and collected separately.
     */
    public static List<PersonalIdCredential> fromJsonArray(JSONArray jsonArray) {
        corruptCredentialItems.clear();
        List<PersonalIdCredential> credentials = new ArrayList<>();
        for (int i = 0; i < jsonArray.length(); i++) {
            JSONObject obj = null;
            try {
                obj = jsonArray.getJSONObject(i);
                PersonalIdCredential credential = new PersonalIdCredential();
                credential.setAppName(obj.getString(META_APP_NAME));
                credential.setSlug(obj.getString(META_SLUG));
                credential.setType(obj.getString(META_TYPE));
                credential.setIssuedDate(obj.getString(META_ISSUED_DATE));
                credential.setTitle(obj.getString(META_TITLE));
                credential.setCredential(obj.getString(META_CREDENTIAL));
                credentials.add(credential);
            } catch (JSONException e) {
                Logger.exception("Error parsing PersonalIdCredential JSON", e);
                handleCorruptCredentialItems(obj);
            }
        }
        return credentials;
    }

    /**
     * Safely handles a corrupt JSON object by attempting to extract whatever is possible.
     */
    private static void handleCorruptCredentialItems(JSONObject obj) {
        if (obj != null) {
            try {
                PersonalIdCredential corruptItem = corruptCredentialItemsFromJson(obj);
                corruptCredentialItems.add(corruptItem);
                Logger.log("corruptCredential", "Corrupt credential parsed with fallback: " + obj);
            } catch (JSONException e) {
                Logger.exception("Failed to parse corrupt PersonalIdCredential", e);
            }
        }
    }

    /**
     * Creates a PersonalIdCredential from corrupt JSON data.
     */
    public static PersonalIdCredential corruptCredentialItemsFromJson(JSONObject json) throws JSONException {
        PersonalIdCredential credItem = new PersonalIdCredential();
        credItem.appName = json.has(META_APP_NAME) ? json.getString(META_APP_NAME) : "";
        credItem.slug = json.has(META_SLUG) ? json.getString(META_SLUG) : "";
        credItem.type = json.has(META_TYPE) ? json.getString(META_TYPE) : "";
        credItem.issuedDate = json.has(META_ISSUED_DATE) ? json.getString(META_ISSUED_DATE) : "";
        credItem.title = json.has(META_TITLE) ? json.getString(META_TITLE) : "";
        credItem.credential = json.has(META_CREDENTIAL) ? json.getString(META_CREDENTIAL) : "";
        return credItem;
    }


    /**
     * Returns all collected corrupt entries, if any.
     */
    public static List<PersonalIdCredential> getCorruptCredentialItems() {
        return new ArrayList<>(corruptCredentialItems);
    }
}

File: app/src/org/commcare/android/database/connect/models/PersonalIdCredential.kt
package org.commcare.android.database.connect.models

import org.commcare.android.storage.framework.Persisted
import org.commcare.models.framework.Persisting
import org.commcare.modern.database.Table
import org.commcare.modern.models.MetaField
import org.javarosa.core.services.Logger
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.Serializable

@Table(PersonalIdCredential.STORAGE_KEY)
class PersonalIdCredential : Persisted(), Serializable {

    @Persisting(1)
    @MetaField(META_UUID)
    var uuid: String = ""

    @Persisting(2)
    @MetaField(META_APP_ID)
    var appId: String = ""

    @Persisting(3)
    @MetaField(META_OPP_ID)
    var oppId: String = ""

    @Persisting(4)
    @MetaField(META_ISSUED_DATE)
    var issuedDate: String = ""

    @Persisting(5)
    @MetaField(META_TITLE)
    var title: String = ""

    @Persisting(6)
    @MetaField(META_ISSUER)
    var issuer: String = ""

    @Persisting(7)
    @MetaField(META_LEVEL)
    var level: String = ""

    @Persisting(8)
    @MetaField(META_TYPE)
    var type: String = ""

    @Persisting(9)
    @MetaField(META_ISSUER_ENVIRONMENT)
    var issuerEnvironment: String = ""

    @Persisting(10)
    @MetaField(META_SLUG)
    var slug: String = ""

    companion object {
        const val STORAGE_KEY = "credential"

        const val META_UUID = "uuid"
        const val META_APP_ID = "app_id"
        const val META_OPP_ID = "opp_id"
        const val META_ISSUED_DATE = "date"
        const val META_TITLE = "title"
        const val META_ISSUER = "issuer"
        const val META_LEVEL = "level"
        const val META_TYPE = "type"
        const val META_ISSUER_ENVIRONMENT = "issuer_environment"
        const val META_SLUG = "slug"

        @JvmStatic
        fun fromJsonArray(jsonArray: JSONArray): List<PersonalIdCredential> {
            val validCredential = mutableListOf<PersonalIdCredential>()
            val corruptCredential = mutableListOf<String>()

            for (i in 0 until jsonArray.length()) {
                var obj: JSONObject? = null
                try {
                    obj = jsonArray.getJSONObject(i)

                    val credential = PersonalIdCredential().apply {
                        uuid = obj.getString(META_UUID)
                        appId = obj.getString(META_APP_ID)
                        oppId = obj.getString(META_OPP_ID)
                        issuedDate = obj.getString(META_ISSUED_DATE)
                        title = obj.getString(META_TITLE)
                        issuer = obj.getString(META_ISSUER)
                        level = obj.getString(META_LEVEL)
                        type = obj.getString(META_TYPE)
                        issuerEnvironment = obj.getString(META_ISSUER_ENVIRONMENT)
                        slug = obj.getString(META_SLUG)
                    }
                    validCredential.add(credential)
                } catch (e: JSONException) {
                    Logger.exception("Corrupt credential at index $i", e)
                    corruptCredential.add("Index $i:\n${obj ?: "Unknown JSON"}\nError: ${e.message}")
                }
            }

            if (corruptCredential.isNotEmpty()) {
                val errorMessage = corruptCredential.joinToString(
                    separator = "\n\n",
                    prefix = "Found ${corruptCredential.size} corrupt credentials:\n"
                )
                Logger.log("CorruptCredentials", errorMessage)
                throw RuntimeException(errorMessage)
            }

            return validCredential
        }

        @JvmStatic
        fun fromV15(old: PersonalIdCredentialV15): PersonalIdCredential {
            val newRecord = PersonalIdCredential()
            newRecord.uuid = old.uuid
            newRecord.appId = old.appId
            newRecord.oppId = old.oppId
            newRecord.issuedDate = old.issuedDate
            newRecord.title = old.title
            newRecord.issuer = old.issuer
            newRecord.level = old.level
            newRecord.type = old.type
            return newRecord
        }
    }
}
